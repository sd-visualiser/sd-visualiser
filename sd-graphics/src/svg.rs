use egui::{Color32, emath::TSTransform};
use sd_core::hypergraph::{
    generic::Ctx,
    traits::{WireType, WithType, WithWeight},
};
use svg::{
    Document, Node,
    node::element::{Circle, Group, Line, Path, Rectangle, Text, path::Data},
};

use crate::{
    common::SCALE,
    shape::{Shape, Shapes},
};

impl<T: Ctx> Shape<T>
where
    <<T as Ctx>::Edge as WithWeight>::Weight: WithType,
{
    pub(crate) fn to_svg(&self) -> Box<dyn Node> {
        match self {
            Self::Operation {
                center,
                radius,
                label,
                ..
            } => {
                let x_size =
                    f32::from(*radius) / 20.0 * f32::from(label.chars().count().max(1) as u16 + 1);
                Box::new(
                    Group::new()
                        .add(
                            Rectangle::new()
                                .set("x", center.x - x_size / 2.0)
                                .set("y", center.y - f32::from(*radius) / 20.0)
                                .set("width", x_size)
                                .set("height", f32::from(*radius) / 10.0)
                                .set("rx", f32::from(*radius) / 20.0)
                                .set("ry", f32::from(*radius) / 20.0)
                                .set("fill", "white")
                                .set("stroke", "black")
                                .set("stroke-width", 1),
                        )
                        .add(
                            Text::new(html_escape::encode_text(label))
                                .set("x", center.x)
                                .set("y", center.y)
                                .set("font-size", 16)
                                .set("font-family", "monospace")
                                .set("text-anchor", "middle")
                                .set("dominant-baseline", "middle"),
                        ),
                )
            }
            Self::CircleFilled {
                center,
                radius,
                addr,
                ..
            } => Box::new(
                Circle::new()
                    .set("cx", center.x)
                    .set("cy", center.y)
                    .set("r", f32::from(*radius) / 20.0)
                    .set(
                        "fill",
                        wire_type_to_svg_colour(addr.weight().get_type()).to_hex(),
                    ),
            ),
            Self::Rectangle { rect, .. } => Box::new(
                Rectangle::new()
                    .set("x", rect.min.x)
                    .set("y", rect.min.y)
                    .set("width", rect.width())
                    .set("height", rect.height())
                    .set("fill", "none")
                    .set("stroke", "gray")
                    .set("stroke-width", 1),
            ),
            Self::Line {
                start, end, addr, ..
            } => Box::new({
                Line::new()
                    .set("x1", start.x)
                    .set("y1", start.y)
                    .set("x2", end.x)
                    .set("y2", end.y)
                    .set(
                        "stroke",
                        wire_type_to_svg_colour(addr.weight().get_type()).to_hex(),
                    )
                    .set("stroke-width", "1")
            }),
            Self::CubicBezier { points, addr, .. } => Box::new({
                let data = Data::new()
                    .move_to((points[0].x, points[0].y))
                    .cubic_curve_to((
                        points[1].x,
                        points[1].y,
                        points[2].x,
                        points[2].y,
                        points[3].x,
                        points[3].y,
                    ));
                Path::new()
                    .set("d", data)
                    .set("fill", "none")
                    .set(
                        "stroke",
                        wire_type_to_svg_colour(addr.weight().get_type()).to_hex(),
                    )
                    .set("stroke-width", 1)
            }),
            Self::Arrow { .. } => {
                panic!("Arrows should not be in svgs")
            }
        }
    }
}

impl<T: Ctx> Shapes<T>
where
    <<T as Ctx>::Edge as WithWeight>::Weight: WithType,
{
    #[must_use]
    pub fn to_svg(&self) -> Document {
        let mut document = Document::new()
            .set("width", self.size.x * SCALE)
            .set("height", self.size.y * SCALE);

        let scale = TSTransform::from_scaling(SCALE);

        for shape in &self.shapes {
            let mut shape = shape.clone();
            shape.apply_tst_transform(&scale);
            document = document.add(shape.to_svg());
        }

        document
    }
}

fn wire_type_to_svg_colour(wire_type: WireType) -> Color32 {
    match wire_type {
        WireType::Data => Color32::BLACK,
        WireType::ControlFlow => Color32::GOLD,
        WireType::SymName => Color32::DARK_GREEN,
        WireType::Colour(colour) => colour,
    }
}
