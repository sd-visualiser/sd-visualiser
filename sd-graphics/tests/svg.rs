use std::collections::HashMap;

use dir_test::{Fixture, dir_test};
use sd_core::{
    common::Direction,
    examples::{self, DummyCtx, DummyEdge, DummyOperation},
    language::{Expr, Language, llvm_ir::parse_llvm_ir, sd_lang::parse_sd_lang},
    lp::Solver,
    monoidal::{
        MonoidalTerm, Slice,
        graph::{MonoidalGraph, MonoidalOp},
        wired_graph::from_graph,
    },
};
use sd_graphics::{layout::layout, render::generate_shapes, shape::Shapes};
use svg::Document;

fn to_svg<T: Language + 'static>(expr: &Expr<T>) -> Document {
    let solver = Solver::default();
    let graph = expr.to_graph(false).unwrap();
    let monoidal_term = from_graph(&graph, solver);
    let monoidal_graph = MonoidalGraph::from(&monoidal_term);
    let layout = layout(&monoidal_graph, solver).unwrap();
    let mut shapes = Vec::new();
    generate_shapes(&mut shapes, &layout, true);
    let shapes = Shapes {
        shapes,
        size: layout.size(),
    };
    shapes.to_svg()
}

#[dir_test(dir: "$CARGO_MANIFEST_DIR/../examples", glob: "**/with*.sd", loader: parse_sd_lang, postfix: "external_inputs")]
#[dir_test(dir: "$CARGO_MANIFEST_DIR/../examples", glob: "llvm-ir/bug-97.ll", loader: parse_llvm_ir)]
fn svg_test<T: Language + 'static>(fixture: Fixture<(&str, Expr<T>)>) {
    let (name, expr) = fixture.content();
    let svg = to_svg(expr);
    insta::assert_binary_snapshot!(
        &format!("{name}.svg"),
        svg.to_string().into(),
        &svg.to_string()
    );
}

#[test]
fn examples() {
    let terms: HashMap<&str, MonoidalTerm<DummyCtx, _>> = [
        (
            "id",
            MonoidalGraph {
                free_inputs: vec![],
                bound_inputs: vec![DummyEdge],
                slices: vec![],
                free_outputs: vec![],
                bound_outputs: vec![DummyEdge],
            },
        ),
        (
            "double_id",
            MonoidalGraph {
                free_inputs: vec![],
                bound_inputs: vec![DummyEdge; 2],
                slices: vec![],
                free_outputs: vec![],
                bound_outputs: vec![DummyEdge; 2],
            },
        ),
        (
            "copy",
            MonoidalGraph {
                free_inputs: vec![],
                bound_inputs: vec![DummyEdge],
                slices: vec![Slice {
                    ops: vec![MonoidalOp::Copy {
                        addr: DummyEdge,
                        copies: 2,
                    }],
                }],
                free_outputs: vec![],
                bound_outputs: vec![DummyEdge; 2],
            },
        ),
        (
            "delete",
            MonoidalGraph {
                free_inputs: vec![],
                bound_inputs: vec![DummyEdge],
                slices: vec![Slice {
                    ops: vec![MonoidalOp::Copy {
                        addr: DummyEdge,
                        copies: 0,
                    }],
                }],
                free_outputs: vec![],
                bound_outputs: vec![],
            },
        ),
        (
            "swap",
            MonoidalGraph {
                free_inputs: vec![],
                bound_inputs: vec![DummyEdge; 2],
                slices: vec![Slice {
                    ops: vec![MonoidalOp::Swap {
                        addrs: vec![
                            (DummyEdge, Direction::Forward),
                            (DummyEdge, Direction::Forward),
                        ],
                        out_to_in: vec![1, 0],
                    }],
                }],
                free_outputs: vec![],
                bound_outputs: vec![DummyEdge; 2],
            },
        ),
        (
            "double_swap",
            MonoidalGraph {
                free_inputs: vec![],
                bound_inputs: vec![DummyEdge; 2],
                slices: vec![
                    Slice {
                        ops: vec![MonoidalOp::Swap {
                            addrs: vec![
                                (DummyEdge, Direction::Forward),
                                (DummyEdge, Direction::Forward),
                            ],
                            out_to_in: vec![1, 0],
                        }],
                    },
                    Slice {
                        ops: vec![MonoidalOp::Swap {
                            addrs: vec![
                                (DummyEdge, Direction::Forward),
                                (DummyEdge, Direction::Forward),
                            ],
                            out_to_in: vec![1, 0],
                        }],
                    },
                ],
                free_outputs: vec![],
                bound_outputs: vec![DummyEdge; 2],
            },
        ),
        ("int", examples::int()),
        ("copy_left_assoc", examples::copy()),
        (
            "copy_right_assoc",
            MonoidalGraph {
                free_inputs: vec![],
                bound_inputs: vec![DummyEdge],
                slices: vec![
                    Slice {
                        ops: vec![MonoidalOp::Copy {
                            addr: DummyEdge,
                            copies: 2,
                        }],
                    },
                    Slice {
                        ops: vec![
                            MonoidalOp::Copy {
                                addr: DummyEdge,
                                copies: 1,
                            },
                            MonoidalOp::Copy {
                                addr: DummyEdge,
                                copies: 2,
                            },
                        ],
                    },
                ],
                free_outputs: vec![],
                bound_outputs: vec![DummyEdge; 3],
            },
        ),
        (
            "copy_left_unit",
            MonoidalGraph {
                free_inputs: vec![],
                bound_inputs: vec![DummyEdge],
                slices: vec![
                    Slice {
                        ops: vec![MonoidalOp::Copy {
                            addr: DummyEdge,
                            copies: 2,
                        }],
                    },
                    Slice {
                        ops: vec![
                            MonoidalOp::Copy {
                                addr: DummyEdge,
                                copies: 0,
                            },
                            MonoidalOp::Copy {
                                addr: DummyEdge,
                                copies: 1,
                            },
                        ],
                    },
                ],
                free_outputs: vec![],
                bound_outputs: vec![DummyEdge; 1],
            },
        ),
        (
            "copy_right_unit",
            MonoidalGraph {
                free_inputs: vec![],
                bound_inputs: vec![DummyEdge],
                slices: vec![
                    Slice {
                        ops: vec![MonoidalOp::Copy {
                            addr: DummyEdge,
                            copies: 2,
                        }],
                    },
                    Slice {
                        ops: vec![
                            MonoidalOp::Copy {
                                addr: DummyEdge,
                                copies: 1,
                            },
                            MonoidalOp::Copy {
                                addr: DummyEdge,
                                copies: 0,
                            },
                        ],
                    },
                ],
                free_outputs: vec![],
                bound_outputs: vec![DummyEdge; 1],
            },
        ),
        ("thunk", examples::thunk()),
    ]
    .into_iter()
    .collect();
    for (name, term) in &terms {
        let layout = layout(term, Solver::default()).unwrap();
        let mut shapes = Vec::new();
        generate_shapes(&mut shapes, &layout, true);
        let shapes = Shapes {
            shapes,
            size: layout.size(),
        };
        let svg = shapes.to_svg();
        insta::assert_binary_snapshot!(
            &format!("{name}.svg"),
            svg.to_string().into(),
            &svg.to_string()
        );
    }
}
