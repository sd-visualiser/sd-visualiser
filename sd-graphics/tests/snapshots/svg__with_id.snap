---
source: sd-graphics/tests/svg.rs
expression: "<svg height=\"175\" width=\"150\" xmlns=\"http://www.w3.org/2000/svg\">\n<line stroke=\"black\" stroke-width=\"1\" x1=\"50\" x2=\"50\" y1=\"0.000000052063797\" y2=\"25\"/>\n<path d=\"M50,25 C50,37.5,50,37.5,50,50\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<rect fill=\"none\" height=\"100\" stroke=\"gray\" stroke-width=\"1\" width=\"100\" x=\"25\" y=\"50\"/>\n<circle cx=\"100\" cy=\"50\" fill=\"black\" r=\"2.5\"/>\n<circle cx=\"100\" cy=\"150\" fill=\"black\" r=\"2.5\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"100\" x2=\"100\" y1=\"50\" y2=\"75\"/>\n<line stroke=\"black\" stroke-width=\"1\" x1=\"100\" x2=\"100\" y1=\"125\" y2=\"150\"/>\n<path d=\"M100,75 C100,95.00001,100,100,100,100\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n<path d=\"M100,100 C100,100,100,104.99999,100,125\" fill=\"none\" stroke=\"black\" stroke-width=\"1\"/>\n</svg>"
extension: svg
snapshot_kind: binary
---
